maven-wrapper/src/main/java/org/apache/maven/wrapper/JdkDownloader.java

```diff
@@ -0,0 +1,127 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.maven.wrapper;
+
+import java.io.IOException;
+import java.net.URI;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+
+/**
+ * Handles downloading and installing JDKs
+ */
+public class JdkDownloader {
+    private final Downloader downloader;
+    private final PathAssembler pathAssembler;
+    private final Verifier verifier;
+
+    public JdkDownloader(Downloader downloader, PathAssembler pathAssembler, Verifier verifier) {
+        this.downloader = downloader;
+        this.pathAssembler = pathAssembler;
+        this.verifier = verifier;
+    }
+
+    public Path download(URI jdkDistribution, String jdkVersion, boolean isToolchain, String sha256Sum)
+            throws Exception {
+        PathAssembler.LocalDistribution localDistribution =
+                getLocalDistribution(jdkDistribution, jdkVersion, isToolchain);
+
+        Path jdkDir = localDistribution.getDistributionDir();
+        Path jdkHome = findJdkHome(jdkDir);
+
+        if (jdkHome != null && Files.exists(jdkHome)) {
+            return jdkHome;
+        }
+
+        Path zipFile = localDistribution.getZipFile();
+
+        boolean downloaded = false;
+        if (!Files.exists(zipFile) || Files.size(zipFile) == 0) {
+            Files.createDirectories(zipFile.getParent());
+            downloader.download(jdkDistribution, zipFile);
+            downloaded = true;
+        }
+
+        // Validate checksum if provided
+        if (sha256Sum != null && !sha256Sum.isEmpty()) {
+            try {
+                verifier.verify(zipFile, "jdkSha256Sum", Verifier.SHA_256_ALGORITHM, sha256Sum);
+            } catch (Exception e) {
+                throw new RuntimeException(
```

Comment:
this should likely be a checked exception since it's outside the control of the program, and definitely shouldn't be a raw RuntimeException.